src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs

```diff
@@ -91,6 +92,27 @@ private TupleTypeDecoder(ImmutableArray<string> elementNames)
             return DecodeTupleTypesInternal(metadataType, elementNames, hasTupleElementNamesAttribute);
         }
 
+        public static TypeSymbolWithAnnotations DecodeTupleTypesIfApplicable(
+            TypeSymbolWithAnnotations metadataType,
+            EntityHandle targetHandle,
+            PEModuleSymbol containingModule)
+        {
+            ImmutableArray<string> elementNames;
+            var hasTupleElementNamesAttribute = containingModule
+                .Module
+                .HasTupleElementNamesAttribute(targetHandle, out elementNames);
+
+            // If we have the TupleElementNamesAttribute, but no names, that's
+            // bad metadata
+            if (hasTupleElementNamesAttribute && elementNames.IsDefaultOrEmpty)
+            {
+                return TypeSymbolWithAnnotations.Create(new UnsupportedMetadataTypeSymbol(), isNullableIfReferenceType: null);
+            }
+
+            var decoded = DecodeTupleTypesInternal(metadataType.TypeSymbol, elementNames, hasTupleElementNamesAttribute);
+            return TypeSymbolWithAnnotations.Create(decoded, isNullableIfReferenceType: metadataType.IsNullable, metadataType.CustomModifiers);
```

Comment:
>return TypeSymbolWithAnnotations.Create(decoded, isNullableIfReferenceType: metadataType.IsNullable, metadataType.CustomModifiers); [](start = 12, length = 131)

It looks like some optimizations won't work properly unless we return the original ```metadataType``` when nothing is changed.